src/nimblepkg/tools.nim

```diff
@@ -29,38 +34,57 @@ proc doCmd*(cmd: string) =
     display("Executing", cmd, priority = MediumPriority)
     let exitCode = execCmd(cmd)
     if exitCode != QuitSuccess:
-      raise newException(NimbleError,
+      raise nimbleError(
         "Execution failed with exit code $1\nCommand: $2" %
         [$exitCode, cmd])
   else:
     displayDebug("Executing", cmd)
     let (output, exitCode) = execCmdEx(cmd)
     displayDebug("Output", output)
     if exitCode != QuitSuccess:
-      raise newException(NimbleError,
+      raise nimbleError(
         "Execution failed with exit code $1\nCommand: $2\nOutput: $3" %
         [$exitCode, cmd, output])
 
-{.warning[Deprecated]: off.}
-proc doCmdEx*(cmd: string): tuple[output: TaintedString, exitCode: int] =
+proc doCmdEx*(cmd: string): ProcessOutput =
   let bin = extractBin(cmd)
   if findExe(bin) == "":
-    raise newException(NimbleError, "'" & bin & "' not in PATH.")
+    raise nimbleError("'" & bin & "' not in PATH.")
   return execCmdEx(cmd)
 
-template cd*(dir: string, body: untyped) =
-  ## Sets the current dir to ``dir``, executes ``body`` and restores the
-  ## previous working dir.
-  let lastDir = getCurrentDir()
-  setCurrentDir(dir)
-  body
-  setCurrentDir(lastDir)
+proc doCmdExAsync*(cmd: string): Future[ProcessOutput] {.async.} =
+  let bin = extractBin(cmd)
+  if findExe(bin) == "":
+    raise nimbleError("'" & bin & "' not in PATH.")
+  let res = await asyncproc.execProcess(cmd)
+  return (res.output, res.exitCode)
```

Comment:
Is `osproc.execProcess` asynchronous?

Reply:
I prefer using the asynctools library because it enables us in the future to implement some parallelization of downloads when doing recursive downloads without using a lock file. This is not possible with the `osproc.execProcesses`.

Reply:
Ok, acceptable if this code finally actually works (on Windows and elsewhere). ;-)